### ISSUE TITLE
Mode 2/3 adjustment

### ISSUE THREAD

**Issue by polosubtil** [Courseplay/courseplay#6065]
Hi again guys,

Just a medium thing here for mode 2/3. Globally, it works pretty well now, especially with the land.
One last thing which causes me some issues sometimes : When the combine is unloading (moving) in a trailer and they hit the end of the row, they stop together and the combine can finish unloading (stopped). Great so far...
First small detail => After finishing unloading, the combine goes forward for a few centimetres and stops again to unload (looks like every 1 %), this repeats until the pipe is no longer on top of the trailer. I can definitely live with that, it works, just losing a bit of time.
Second detail => a lot of my fields have crazy shapes (Oakfield Farm), so sometimes the combine is actually "stuck" because consider a vehicle is in the way.

So my question would be => Could we apply a reverse driving for the trailer for a few meters when the unload finishes at the end of a row ? Exactly what we do in the first headland ??? That would solve both situations.

Last thing, the "Drive now button" doesn't work anymore in Mode 2, I can give you a log if you tel me which debug you want me to activate. Not major again, just really useful when your trailer is almost full but didn't reach the level you set to drive the course.

![fsScreen_2020_08_25_10_24_31](https://user-images.githubusercontent.com/60451845/93005172-d4859580-f54e-11ea-973a-9a158b655daf.png)
![fsScreen_2020_09_12_23_20_15](https://user-images.githubusercontent.com/60451845/93005175-dc453a00-f54e-11ea-8dbd-b31c6a7e03f5.png)

**Comment by polosubtil** [comment]
Just to add a quick idea here
In case the trailer is full while unloading at the end of the row (UNLOADING_STOPPED_COMBINE), the combine isn't empty and usually they both stay stuck where they are. Os I guess in that case the tractor still needs to reverse and wait for the combine to finish its maneuver before starting to drive the unload course to the silo

**Comment by pvaiko** [comment]
> So my question would be => Could we apply a reverse driving for the trailer for a few meters when the unload finishes at the end of a row ? Exactly what we do in the first headland ??? That would solve both situations.

Yes we could, my only problem is that testing Mode 2 is real pain in the butt if you'd be willing to test a branch before I push it to master that'd help a lot.

> In case the trailer is full while unloading at the end of the row (UNLOADING_STOPPED_COMBINE), the combine isn't empty and usually they both stay stuck where they are.

For that I'd need at least a log, I have difficulties to set up this scenario.

**Comment by pvaiko** [comment]
> First small detail => After finishing unloading, the combine goes forward for a few centimetres and stops again to unload (looks like every 1 %), this repeats until the pipe is no longer on top of the trailer. I can definitely live with that, it works, just losing a bit of time.

For this, we could also change the combine's behavior to not attempt unloading again in this case.

**Comment by polosubtil** [comment]
Hi @pvaiko 
No problem, happy to test for you
Let me start with the log, which channel do you need ? I will set it up tonight

**Comment by pvaiko** [comment]
7, 12 and 14

**Comment by polosubtil** [comment]
`For this, we could also change the combine's behavior to not attempt unloading again in this case. `
Yes it would work, I just thought reversing the trailer would solve both things

**Comment by polosubtil** [comment]
Hi @pvaiko 
To start with, just a log (Channel 7, 12, 14) when the trailer isn't full but the combine struggle to get away as it keeps stopping to unload 1%. So in that case, the only "problem is it takes time to get moving, but it works. I am trying to get a log of the situation where the trailer is empty
[log.txt](https://github.com/Courseplay/courseplay/files/5214985/log.txt)
![fsScreen_2020_09_13_21_26_03](https://user-images.githubusercontent.com/60451845/93026685-348c4280-f608-11ea-9f30-20f5b37bf3d6.png)
![fsScreen_2020_09_13_21_27_13](https://user-images.githubusercontent.com/60451845/93026690-3fdf6e00-f608-11ea-8099-8bf018bd56e5.png)
![fsScreen_2020_09_13_21_27_15](https://user-images.githubusercontent.com/60451845/93026696-466de580-f608-11ea-897f-83d390792b4a.png)

**Comment by polosubtil** [comment]
Here it is, finally. A real pain to get ;)
In that configuration, the combine stopped at the end of the row, kept unloading in the trailer until it was full. So now the combine is not empty and the trailer is full, and nobody is moving.
As I already said, unfortunately the "drive now" button doesn't work anymore so I have to stop the unloader course, the combine then do the turn, and I restart the unloader course just after the combine finish its turn so it can drive the course to unload the trailer in the silo.
[log.txt](https://github.com/Courseplay/courseplay/files/5215040/log.txt)
![fsScreen_2020_09_13_22_19_44](https://user-images.githubusercontent.com/60451845/93027738-576e2500-f60f-11ea-84a8-c20817030b40.png)

**Comment by rohne83** [comment]
Recently experienced the same situations even when the combine could get emptied totally. But not sure what exact setup it was. Guess it was also when combine wanted to turn left after the row while the trailer's course would cross the combine's (i.e. turning sharp right). Always had to fire the trailer's helper to resolve the situation. But really hard to get such situation reproduced.

**Comment by pvaiko** [comment]
@polosubtil I pushed a change on the Mode2 branch if you want to try.

**Comment by polosubtil** [comment]
Hi @pvaiko 
I tried your version. I can't reproduce it now ;)
Basically, every time I get to the end of the row, the tractor stops a few meter head, let the combine do its turn and come back to the combine to finish the unloading. I saw that yesterday to get the log file I gave you but I manage to make it work. I wasn't able this time.
[log.txt](https://github.com/Courseplay/courseplay/files/5219884/log.txt)
I did it 3 times in a row in the attached log. I think this isn't what you planned but that works too ;)

Here the tractor stops right before the end of the row
![fsScreen_2020_09_14_18_58_22](https://user-images.githubusercontent.com/60451845/93115678-7bd40b00-f6bc-11ea-8d7b-4659aa075b4f.png)
After the combine finishes its turn, the tractor comes back to continue the unload
![fsScreen_2020_09_14_18_58_53](https://user-images.githubusercontent.com/60451845/93115691-81315580-f6bc-11ea-8c64-d1d3a69c0e3c.png)

**Comment by pvaiko** [comment]
>  I think this isn't what you planned but that works too ;)

LOL, true, but still kind of cool, isn't it?

So you say the unloading began before the end of the row, but this time the combine did not stop until it was empty (or the trailer full), but continued to turn?

**Comment by polosubtil** [comment]
Real cool man ;)

Yes the unloading starts a little bit before the end of the row (maybe 25/30 m), the tractor stops just before the end of the row and the combine doesn't stop at all

**Comment by pvaiko** [comment]
Ok, I think I understand why is this happening, made another push. See, this is why Mode 2 is difficult, with my tractor/combine this did not happen at all...

**Comment by polosubtil** [comment]
We are back to normal, the combine and trailer both stop at the end of the row but when the trailer is full and the combine not empty nothing happened, they stay like that and wait
[log.txt](https://github.com/Courseplay/courseplay/files/5220187/log.txt)

Start unloading while moving
![fsScreen_2020_09_14_20_12_33](https://user-images.githubusercontent.com/60451845/93122599-10436b00-f6c7-11ea-9e53-f5b254958e9f.png)

The combine stops at the end of the row. The tractor too
![fsScreen_2020_09_14_20_12_53](https://user-images.githubusercontent.com/60451845/93122603-13d6f200-f6c7-11ea-9c8d-d36db47c33ec.png)

Combine isn't empty and trailer is full => they stay like that
![fsScreen_2020_09_14_20_13_05](https://user-images.githubusercontent.com/60451845/93122611-176a7900-f6c7-11ea-8f4a-5685b9124938.png)

**Comment by polosubtil** [comment]
If the trailer is NOT full, it works like a charm !!!!!!!!!!!!!!!
See end of the log file only (sorry, didn't restart the game)
[log.txt](https://github.com/Courseplay/courseplay/files/5220251/log.txt)

Start unloading while moving
![fsScreen_2020_09_14_20_24_22](https://user-images.githubusercontent.com/60451845/93123512-885e6080-f6c8-11ea-9ed1-d49e47362282.png)

Both stop to finish unloading
![fsScreen_2020_09_14_20_24_31](https://user-images.githubusercontent.com/60451845/93123537-8eecd800-f6c8-11ea-85c5-09f88c92c135.png)

Tractor reverse once the combine is empty
![fsScreen_2020_09_14_20_24_40](https://user-images.githubusercontent.com/60451845/93123549-957b4f80-f6c8-11ea-8bd2-ed557ff5d375.png)

Combine gets ready to move
![fsScreen_2020_09_14_20_24_42](https://user-images.githubusercontent.com/60451845/93123566-9c09c700-f6c8-11ea-887a-6edd02cfbe98.png)

Combine finishes the row and starts its turn maneuver 
![fsScreen_2020_09_14_20_24_46](https://user-images.githubusercontent.com/60451845/93123577-a1ffa800-f6c8-11ea-9d35-3db5dd046422.png)

**Comment by pvaiko** [comment]
Ok, this first iteration addressed only the combine empty part. Next comes the trailer full part.

**Comment by polosubtil** [comment]
Ok ! So great job then !!!!!!!
My save game is ready for the next test ;), just shoot whenever you're ready

**Comment by rohne83** [comment]
This looks quite cool, guys! Will this behavior be applicable to mode 3 overloader as well, once finished and merged to the main branch?

**Comment by pvaiko** [comment]
yes, Mode 3 is the same as Mode 2 except unload course to the silo.

**Comment by pvaiko** [comment]
Pushed another change, check it out if you like, this is for the full trailer.

Make sure you have debug **channel 4** activated as well, I forgot it last time.

**Comment by polosubtil** [comment]
Amazing work, thanks a lot @pvaiko !!!! You rock ;)
It works like a charm : Combine continue unloading in the trailer while stopped at the end of the row. As soon as the trailer is full (combine not empty), the tractor reverse, the combine turns and the tractor goes to the unload course once the combine finish turning => Perfect !!!
[log.txt](https://github.com/Courseplay/courseplay/files/5226747/log.txt)
Logfile of the situation if you want to check for errors

Finish filling up the trailer
![fsScreen_2020_09_15_19_40_37](https://user-images.githubusercontent.com/60451845/93245524-c1133e00-f78b-11ea-8c0a-a8147035f536.png)

Tractor reverse once full
![fsScreen_2020_09_15_19_40_48](https://user-images.githubusercontent.com/60451845/93245561-c83a4c00-f78b-11ea-8099-38c46549aa3f.png)
![fsScreen_2020_09_15_19_40_50](https://user-images.githubusercontent.com/60451845/93245586-ce302d00-f78b-11ea-8f70-311dd6d10823.png)

Tractor waits for combine to turn
![fsScreen_2020_09_15_19_40_53](https://user-images.githubusercontent.com/60451845/93245607-d4bea480-f78b-11ea-9441-f03125b7cd69.png)

Tractor goes to the unload course once the combine finishes its turn
![fsScreen_2020_09_15_19_41_14](https://user-images.githubusercontent.com/60451845/93245620-db4d1c00-f78b-11ea-9fd1-bcf416c94e2e.png)

**Comment by polosubtil** [comment]
Damn !!! Let me try again with Channel 4

**Comment by pvaiko** [comment]
If it is working I don't need a log :)

**Comment by polosubtil** [comment]
It's working for me, looks really good to me
